stable/metallb: Refactor the config-passing part of the chart.

danderson opened the pull request:
Replaces `simpleLayer2` with a choice between 2 ways of configuring MetalLB:

- use `configInline` to specify a MetalLB configuration in values.yaml. Helm will manage the ConfigMap in the release.
- use `configName` to specify an existing external ConfigMap. Helm will not manage the ConfigMap, it will just point to it in pod specs.

danderson committed:
stable/metallb: Refactor the config-passing part of the chart.

Adds a third option for configuration: passing in a reference to
an external (not managed by Helm) ConfigMap. This makes life easier
for some users who want Helm to manage the binary lifecycle, but
want to keep the configuration manually editable without special tools.

The 3 ways of passing a config are now grouped under the 'config' yaml
tree, to make things more readable.

unguiculus commented on stable/metallb/values.yaml:
```diff
@@ -2,16 +2,21 @@
+  # simpleLayer2 accepts an IP address range and generates a trivial
+  # "quick start" MetalLB configuration that uses layer 2 mode. Refer
+  # to https://metallb.universe.tf/ for more information. This option
+  # is not recommended for production use.
```
I don't understand this. In the PR description you mention this is an existing config map, but the comment above suggests it's an IP address. Can you clarify this?

danderson replied:
The objective of this PR is to give users 3 alternative ways to configure MetalLB:

- If you provide a name in `config.name`, we'll use that existing ConfigMap.
- If you put YAML under `config.inline`, Helm will create the ConfigMap with that YAML.
- If you put IP addresses in `config.simpleLayer2`, Helm will generate a "quick start" config for those IPs.

You have to choose one of the three options, and the chart adjusts for each one.

Do you have a suggestion for how I can make this more understandable in the file comments?

danderson replied:
After thinking about it more, I've removed simpleLayer2 completely, because it was encouraging the wrong behavior from users. There are now 2 options for how to configure MetalLB. I've updated the PR description and docstrings in values.yaml.

danderson committed:
Remove simpleLayer2 as an option.

It complicates the chart, and it's not really a great option anyway
because it encourages using MetalLB without reading the documentation
first, which leads to broken clusters.

Since we're now back down to 2 options, switch back to non-nested configs.

danderson commented:
Updated, please take another look?

danderson commented:
Ping. I'm about to release MetalLB 0.6, and this PR is a prerequisite for updating the helm chart to this new release. If there's any way I can make this review easier for you, please tell me :)

/assign @mattfarina

unguiculus commented on stable/metallb/templates/config.yaml:
```diff
@@ -10,13 +10,5 @@ metadata:
-      cidr:
-      - {{ .Values.arpAddresses }}
-{{- end }}
+{{ toYaml .Values.configInline | indent 4 }}
```
Could you change this to `existingConfigMap` to be more in line with other charts (`existingConfigMap`, `existingClaim`, `existingSecret`, ...)

danderson replied:
Done.

unguiculus commented:
/ok-to-test

danderson committed:
Rename configName to existingConfigMap to align with other charts.

danderson commented:
Changed as requested. Thanks for the review!

unguiculus commented:
/lgtm
